doc/neps/nep-0049.rst

```diff
@@ -162,8 +162,16 @@ the ``sz`` argument is correct.
     #include <numpy/arrayobject.h>
     NPY_NO_EXPORT void *
 
-    shift_alloc(size_t sz) {
-        char *real = (char *)malloc(sz + 64);
+    typedef struct {
+        void *(*malloc)(size_t);
+        void *(*calloc)(size_t, size_t);
+        void *(*realloc)(void *, size_t);
+        void (*free)(void *);
+    } Allocator;
```

Comment:
The idea in this example is that `Allocator` is some third-party allocator API that a user is gluing to numpy, I assume?

This example confuses me a little because it does two orthogonal things:

* wraps some `Allocator` API that is not part of this spec
* implements some shifting logic

It might be clearer as two separate examples:

* One which wraps some imaginary third-party allocator API into a numpy memory handler
* One which takes any existing numpy memory handler and does this shifting logic

For the latter, a more naturalchoice of type for `ctx` would be something like
```C
struct {
    PyDataMem_Handler inner_handler;
    size_t offset;
} shift_context_t;
```
which avoids introducing a second allocator API `Allocator` which isn't the one the PR is talking about

Reply:
> The idea in this example is that `Allocator` is some third-party allocator API that a user is gluing to numpy, I assume?

That's exactly the concept, which is, I believe,  a fair use of the `ctx` functionality. Good both for testing and demonstration.

> It might be clearer as two separate examples:
> 
>     * One which wraps some imaginary third-party allocator API into a numpy memory handler
> 
>     * One which takes any existing numpy memory handler and does this shifting logic

I agree. It probably would make more sense to split the example.

> For the latter, a more naturalchoice of type for `ctx` would be something like
> 
> ```c
> struct {
>     PyDataMem_Handler inner_handler;
>     size_t offset;
> } shift_context_t;
> ```
> 
> which avoids introducing a second allocator API `Allocator` which isn't the one the NEP is talking about

I guess the `shift` example could not make use of a `ctx` at all (NULL).

Reply:
> I guess the shift example could not make use of a ctx at all (NULL).

Only if it hard-codes the shift amount as a compile-time constant. I think it's better to have it use the ctx because it makes it clear how it can be used even in simple cases.